### ISSUE TITLE
Support for generated columns in postgres and mysql

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#4894]
Add support for specifying a column as generated by the database with `GENERATED ALWAYS`

Currently postgres and mysql support it (the list may not be complete)
https://www.postgresql.org/docs/12/ddl-generated-columns.html
https://dev.mysql.com/doc/refman/8.0/en/create-table-generated-columns.html

I've tried to search in the documentation and it does not seems to be a way to specify a generated column in sqlalchemy. 
I've found a stack overflow answer https://stackoverflow.com/questions/57596603/sqlalchemy-mysql-column-ddl-with-generated-always-as but it does not seem to be the cleanest way to deal with it.

maybe a Generated construct used in the column with the code would be appropriate?
Something like
```py
sa.Table('foo', meta,
  sa.Column('quantity', sa.Integer, nullable=False),
  sa.Column('price', sa.Float, nullable=False),
  sa.Column('revenue', sa.Float, Computed('quantity * price', always=True))
)
```

edit: updated example following https://github.com/sqlalchemy/sqlalchemy/issues/4894#issuecomment-539050154

Current support and syntax seems to be:

- Postgress: supports only stored 
`GENERATED ALWAYS AS ( generation_expr ) STORED`
  https://www.postgresql.org/docs/12/ddl-generated-columns.html
  https://www.postgresql.org/docs/12/sql-createtable.html

- mysql | mariad db:  supports both stored and virtual
  `col_name data_type [GENERATED ALWAYS] AS (expr) [VIRTUAL | STORED]`
  (maridadb supports also `PERSISTENT`, but is an alias to `STORED` 
https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html
https://mariadb.com/kb/en/library/generated-columns/
- mssql: support both stored and virtual
`<computed_column_definition> ::=
column_name AS computed_column_expression
[ PERSISTED [ NOT NULL ] ]`
https://docs.microsoft.com/en-us/sql/t-sql/statements/create-table-transact-sql?view=sql-server-2017
- oracle: support only virtual 
`column_name [datatype] [GENERATED ALWAYS] AS (expression) [VIRTUAL]`
https://oracle-base.com/articles/11g/virtual-columns-11gr1

**Comment by zzzeek** [comment]
there's not and I would need PR contributions for this to happen.

is GENERATED part of the SQL standard?  if not, then the keyword arguments need to be mysql_generated, postgresql_generated

**Comment by zzzeek** [comment]
if this is more like a SQL standard I might favor a real Generated construct that delivers keywords like ALWAYS and STORED based on flags

**Comment by zzzeek** [comment]
e.g. it works like CheckConstraint a bit

**Comment by CaselIT** [comment]
From wikipedia (hardly the best source) it seems that is not sql standard
https://en.wikipedia.org/wiki/Virtual_column

**Comment by zzzeek** [comment]
so mysql and PG are..using the same syntax ?   did one get it from the other?

**Comment by zzzeek** [comment]
well here's oracle, this is looking like a standard syntax:
```

CREATE TABLE employees (
  id          NUMBER,
  first_name  VARCHAR2(10),
  last_name   VARCHAR2(10),
  salary      NUMBER(9,2),
  comm1       NUMBER(3),
  comm2       NUMBER(3),
  salary1     AS (ROUND(salary*(1+comm1/100),2)),
  salary2     NUMBER GENERATED ALWAYS AS (ROUND(salary*(1+comm2/100),2)) VIRTUAL,
  CONSTRAINT employees_pk PRIMARY KEY (id)
);
```

**Comment by zzzeek** [comment]
SQL Server, different, but not terrible

```
CREATE TABLE dbo.Products
   (
      ProductID int IDENTITY (1,1) NOT NULL
      , QtyAvailable smallint
      , UnitPrice money
      , InventoryValue AS QtyAvailable * UnitPrice
    )
;
```

**Comment by CaselIT** [comment]
> if this is more like a SQL standard I might favor a real Generated construct that delivers keywords like ALWAYS and STORED based on flags

Yes that is probably best since there seems to be stored/virtual option (and maybe other)

I'll update the initial issue

**Comment by zzzeek** [comment]
I'd like a new construct for this..

officially, it would be Virtual(), but that seems a little awkward.  Computed() seems a lot better.

Behaviorally, the construct acts a lot like a server_onupdate, except it accommodate for server_default too.   this has to be worked out.  when this construct is present, both of those have to be implicitly set up, as this impacts the ORM knowing it has to fetch these (e.g. server_onupdate=FetchedValue()).

Syntax would be like a CheckConstraint() to make it somewhat succinct:

```
class Model(Base):

    id = Column(Integer, primary_key=True)
    quantity = Column(Integer, nullable=False)
    unit_price_taxed = Column(DECIMAL(20, 4), nullable=False)
    line_price_taxed = Column(
        DECIMAL(20, 4),
        Computed("quantity * unit_price_taxed", always=True),
        index=True,
    )

```

above, it would raise an error to also set server_default or server_onupdate because Computed is present.   sqlalchemy.schema.Computed would extend sqlalchemy.schema.FetchedValue and would be automatically assigned to server_default and server_onupdate when used.

tests go into test/sql/test_defaults.py, or alternatively a new suite test/sql/test_computed.py.   these should be round trip tests for those databases that support comptued columns.   then each backend dialect gets compilation tests.

dialects to support first, Postgresql and MySQL, and very hopefully Oracle and MSSQL.   I can accept a PR for just the first two.   For Firebird, I don't really care, that's mostly a not-supported dialect right now.    This is enough of a feature that I can take over a PR and finish it up if someone wants to just put some effort into getting it started.

I didn't realize that many DBs support this by now so this is actually kind of a high priority.   if there's no authors for a PR I might find time to get around to this alone.

**Comment by CaselIT** [comment]
`Computed("quantity * unit_price_taxed", always=True),`

I'm not sure if `always` if best to convey if they should be stored or not.
maybe `stored` would be cleared (from what I understand a stored computed column is computed on insert/update, while a not stored one is computed in the select, like in a normal view)

**Comment by zzzeek** [comment]
Sure, basically we need a matrix of what behaviors are selectable among these four vendors and then come up with flags  + sensibele defaults

**Comment by zzzeek** [comment]
oracle: https://oracle-base.com/articles/11g/virtual-columns-11gr1

https://docs.microsoft.com/en-us/sql/relational-databases/tables/specify-computed-columns-in-a-table?view=sql-server-2017

**Comment by CaselIT** [comment]
seems to be:
- Postgress: `STORED` is required: [PostgreSQL currently implements only stored generated columns.](https://www.postgresql.org/docs/12/ddl-generated-columns.html)
- mysql | mariad db: `[VIRTUAL | STORED]` (maridadb supports also `PERSISTENT`, but is an alias to `STORED`
- mssql: `PERSISTED` is optional
- oracle: seems to not support stored

**Comment by zzzeek** [comment]
oracle: "If the datatype is omitted, it is determined based on the result of the expression. The GENERATED ALWAYS and VIRTUAL keywords are provided for clarity only."  e.g. they dont support "persistent" .   SQL Server seems to support "persisted" as an option

**Comment by zzzeek** [comment]
so what is the word "ALWAYS"?   that's just boilerplate?

**Comment by CaselIT** [comment]
> so what is the word "ALWAYS"? that's just boilerplate?

not sure. bot generated and always are optional in oracle, while seem required in PG

**Comment by zzzeek** [comment]
oh and PG / Oracle have mutually exclusive storage implementations.   So we likely should support that too, for people who don't care if its stored or not based on backend, some kind of "auto" option

**Comment by zzzeek** [comment]
OK so it's basically like what MySQL says: "AS (expr) indicates that the column is generated and defines the expression used to compute column values. AS may be preceded by GENERATED ALWAYS to make the generated nature of the column more explicit."    no biggie

**Comment by CaselIT** [comment]
also in mysql maria db `generated always` seems to optional

**Comment by zzzeek** [comment]
weird this is not SQL standard?  it is sure acting like it

**Comment by CaselIT** [comment]
I only found the wikipedia article, so it may be in the sql standard (or proposed)

**Comment by zzzeek** [comment]
welp they say it's not part of any standard, that is enough evidence for me ( I have the one version of the stnadard here and it's a PITA to search)

**Comment by CaselIT** [comment]
I'll update the initial comment with the links and current support

**Comment by CaselIT** [comment]
I've update the initial comment.

> Sure, basically we need a matrix of what behaviors are selectable among these four vendors and then come up with flags + sensibele defaults

I think the default for all other than PG is to not specify anything and use the default selection of the engine.
Postgres requies `STORED` so the is no option

I'm not sure if having two flag `stored` and `virtual` is best or if something like `storage_mode='auto'|'stored'|'virtual'|'persisted'|'persistent'` would be more appropriate

**Comment by zzzeek** [comment]
there's basically three options I can see:  persisted, non-persisted, and backend default.   if you want your model to work on Oracle and Postgresql you would not set this flag, meaning, use backend default.   So I would propose `persisted=True|False|None` with `None` as the default.

another use case, which I dont think I want to do, is when someone's model is to support Oracle *and* mysql, but only on mysql they want the stored version.

**Comment by CaselIT** [comment]
Can't they do `persisted=isMysql()` to support that use case?

**Comment by zzzeek** [comment]
yes they can do those things but most sqlalchemy applications set up their class/table metadata at the module import level and their database in use at runtime via the URL that comes from config. that's why we have to have things like `Integer().with_variant("mysql", TinyInt())`.

**Comment by CaselIT** [comment]
Oh ok. I usually have the config while importing the module, that's why I suggested that option.

Maybe since PG and Oracle have only one option, they could ignore the flag? (Maybe raising a warning?)
It's not a great implementation though.

**Comment by zzzeek** [comment]
I think for now, we can go with that if they specified "None", we do whatever it default for that database.  if someone truly wants to tune this back and forth on different backends there are ways to use events to make that happen at runtime.

**Comment by CaselIT** [comment]
Sorry, wrong button.

Should sqlalchemy raise an error in case of incompatible option in PG and Oracle, or let the database raise it?

I'll try to give this a go, but I'm not familiar with sqlalchemy internals, so I'm promising anything.

**Comment by zzzeek** [comment]
we're going to have compilation rules in each dialect to render these so we have to raise CompileError in the dialects if an option is not supported, becuase we don't know what to render otherwise.

**Comment by CaselIT** [comment]
According to this https://www.postgresql.org/about/news/1976/

>PostgreSQL 12 introduces "generated columns." Defined in the SQL standard, this type of column computes its value from the contents of other columns in the same table. In this version, PostgreSQL supports "stored generated columns," where the computed value is stored on the disk.

they do seem to be in the standard. There is no link for it though

**Comment by zzzeek** [comment]
so SQL 2003 includes "GENERATED AS" in the spec specifically for identity columns.   so it's in the spec, but the Wikipedia entry refers to "virtual columns" which aren't really stored, so *that's* the part that's not in the spec.


anyway we are good to go on this thanks for being involved!



```
<column definition> ::=
<column name> [ <data type or domain name> ]
[ <default clause> | <identity column specification> | <generation clause> ]
[ <column constraint definition>... ]
[ <collate clause> ]
<data type or domain name> ::=
<data type>
| <domain name>
<column constraint definition> ::=
[ <constraint name definition> ] <column constraint> [ <constraint characteristics> ]
<column constraint> ::=
NOT NULL
| <unique specification>
| <references specification>
| <check constraint definition>
<identity column specification> ::=
GENERATED { ALWAYS | BY DEFAULT } AS IDENTITY
[ <left paren> <common sequence generator options> <right paren> ]
<generation clause> ::= <generation rule> AS <generation expression>
<generation rule> ::= GENERATED ALWAYS
<generation expression> ::= <left paren> <value expression> <right paren>
```

**Comment by CaselIT** [comment]
It seems that mysql supports not null with a generated column while mariadb does not

should I add checks in the dialect or demand them to the db?

**Comment by CaselIT** [comment]
mssql does not accept the type on a computed colum. does not allow null but allows not null on persisted columns.

**Comment by zzzeek** [comment]
mssql's type is likely..derived from the expression?   I think for MSSQL we just omit the type when we render.   for null/not null, again if we can get by without rendering anything for "null" / "not null" by default, that would be best.   

I saw you mention you were calling it "Computed".    Should we call it "Generated"?   since that's what the SQL looks like, it might be clearest.

**Comment by CaselIT** [comment]
already renamed in the pr. I can revert it
for me its equally clear. maybe generated is more like generate of sql?

regarding the not null there the default (ie omitting the nullable should always work)

**Comment by zzzeek** [comment]
it's a hard call because Computed describes it better but Generated is more what the SQL looks like.   just leave it for now whichever way you have it

**Comment by CaselIT** [comment]
> mssql's type is likely..derived from the expression? I think for MSSQL we just omit the type when we render

Yes it infers it. But it is strange that it errors of the type is specified. Oracle also infers it but does not error if the type is specified (maybe if it is not compatible?).
I have omitted the type in mssql when the colum is computed

**Comment by CaselIT** [comment]
I've done a fist implementation. I'm not sure if the inspect module needs to be updated or how to update it.

**Comment by sqla-tester** [comment]
**CaselIT** has proposed a fix for this issue in the **master** branch:

**Support for generated columns** https://gerrit.sqlalchemy.org/1539

**Comment by CaselIT** [comment]
I'm not sure if we want to close this after #4928 is merged or if it should stay open since the orm support is still missing

**Comment by zzzeek** [comment]
I'm adding the ORM test, that just tests that the unit of work recognizes the Computed as a server / on update default, to the current gerrit.    The only thing I see beyond that is the flag that gets into "this is a read only attribute" and I think as that is more of a behavioral change that should also impact any expression level column_property(), it should likely be a separate issue and maybe targeted at 1.4.    the basic ability for you to use computed columns including that the ORM does the right thing (except for if you write to the column in Python) will be there in 1.3.x with the current PR.

**Comment by CaselIT** [comment]
I started something regarding the read only attributes in #4896 if you want to take a look.

I don't remember if I also added orm tests there.

I agree that it could be done in a separate issue

**Comment by zzzeek** [comment]
well next bad news, Oracle RETURNING does not work for computed columns.    easiest approach right now is add another supports to the tests and eager_defaults with Oracle  + Computed just won't work.

**Comment by zzzeek** [comment]
though need to see what PG does

**Comment by zzzeek** [comment]
pg 12 works.  OK

**Comment by CaselIT** [comment]
I guess reverting oracle to the MySQL behavior in case of computed columns would not be simple

**Comment by zzzeek** [comment]
oracle is really bad. the RETURNING returns the value *before* the UPDATE happens.  so it's not only wrong it's misleading.  im having it emit a warning.

**Comment by CaselIT** [comment]
that seems like it was implemented to intentionally cause problems!

**Comment by zzzeek** [comment]
the latest gerrit has a lot of oracle additions if you want to look.l

**Comment by CaselIT** [comment]
Nice, since the orm test from the #4896 are in this issue, I think we can probably close that and open a new pr with only the read only attributes parts in it

**Comment by sqla-tester** [comment]
**CaselIT** has proposed a fix for this issue in the **rel_1_3** branch:

**Support for generated columns** https://gerrit.sqlalchemy.org/1559

**Comment by jcoelho93** [comment]
Does the automap extension also support generated columns?
From my tests it does not... If I automap a table and then try to insert a row I get the following error:

```python
sqlalchemy.orm.exc.FlushError: Instance <users at 0x7f5c256010d0> has a NULL identity key.  If this is an auto-generated value, check that the database table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush() is not occurring at an inappropriate time, such as within a load() event.
```

My test:
```python
...
metadata = MetaData(schema='db')
metadata.reflect(engine, only=['users'])
Base = automap_base(metadata=metadata)
Base.prepare()
Users = Base.classes.users

user = User(
    # id=1 wether I pass in the id or not it fails
    name='jdoe'
)
session.add(user)
session.commit()
```

**Comment by nburns** [comment]
For anyone stumbling across this (was one of the first search results for me):

`(MySQLdb.OperationalError) (3105, "The value specified for generated column 'column_name' in table 'table' is not allowed.")`

I was able to map the column in a declarative model by specifying a `server_default` of the column name itself:
```
    time_created_datestamp: Mapped[Optional[int]] = mapped_column(server_default="time_created_datestamp")
```

`time_created_datestamp` is a generated stored column in my database

**Comment by CaselIT** [comment]
hi, you can simply use https://docs.sqlalchemy.org/en/20/core/defaults.html#sqlalchemy.schema.FetchedValue to indicate that the orm should get the value of that column from the db.
adding a server default has a similar effect, so that's why it works the same
